### PR TITLE
IN-943 Finance Order transformation

### PR DESCRIPTION
**Pull request by devkokov**
## Purpose

IN-943

## Approach

Create a finance order in the finance_order table for every Active and Closed order of type 1, 2, 40 and 41 that has a risk level set in order."Ord Risk Lvl". Join the risk_assessment table on "Order No" and use the earliest risk_assessment."Start Date" as billing_start_date for the finance order.

## Checklist

* [ ] I have performed a self-review of my own code
* [ ] I have done an adhoc run against preprod (only needed for high complexity PRs)
* [ ] I have added relevant logging with appropriate levels to my code
* [ ] I have updated documentation where relevant
* [ ] I have added tests to prove my work
* [ ] The product team have tested these changes

### CODE REVIEW

File: migration_steps/transform_casrec/transform/app/entities/finance_order/__init__.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+def runner(target_db, db_config):
+    """
+    | Name          | Running Order | Requires |
+    | --------------| ------------- | -------- |
+    | finance_order | 1             |          |
+    """
+
+    entity_name = "finance_order"
+    extra_entities = []
```
Comment (townxelliot) [inline migration_steps/transform_casrec/transform/app/entities/finance_order/__init__.py:17]: Does this have no dependencies at all? This check can be removed if not.

File: migration_steps/transform_casrec/transform/app/entities/finance_order/finance_order.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        f'''
+            SET datestyle = "ISO, DMY";
+            SELECT MIN("Start Date"::date) AS billing_start_date,
+            "Order No" AS c_order_no
+            FROM {db_config["source_schema"]}.risk_assessment
+            GROUP BY "Order No";
+        '''
+    )
+    risk_assessment_df = pd.read_sql_query(risk_assessment_query, db_config["db_connection_string"])
+    risk_assessment_df = risk_assessment_df[["billing_start_date", "c_order_no"]]
```
Comment (townxelliot) [inline migration_steps/transform_casrec/transform/app/entities/finance_order/finance_order.py:35]: This line is not necessary as these are the only two fields in the dataframe anyway.

**Review by townxelliot — CHANGES_REQUESTED** [review]

**Comment by townxelliot** [comment]
NB my PR for IN-1061 includes the same code as you have to fall back to table definitions if there is no casrec mapping for a column.

File: migration_steps/transform_casrec/transform/app/entities/finance_order/__init__.py (hunk 1 above)
Comment (devkokov) [reply migration_steps/transform_casrec/transform/app/entities/finance_order/__init__.py:17]: Good point, it does depend on cases

File: migration_steps/transform_casrec/transform/app/entities/finance_order/finance_order.py (hunk 2 above)
Comment (devkokov) [reply migration_steps/transform_casrec/transform/app/entities/finance_order/finance_order.py:35]: True 🤦

File: migration_steps/transform_casrec/transform/app/utilities/basic_data_table.py [hunk 3]
```diff
@@ -21,6 +21,8 @@ def get_source_table(mapping_dict):
     no_dupes = list(set(source_table_list))
     if len(no_dupes) == 1:
         return list(set(source_table_list))[0]
+    elif len(no_dupes) == 0:
```
Comment (townxelliot) [inline migration_steps/transform_casrec/transform/app/utilities/basic_data_table.py:24]: NB my PR for IN-1061 includes the same code as you have to fall back to table definitions if there is no casrec mapping for a column: https://github.com/ministryofjustice/opg-data-casrec-migration/pull/491/files#diff-1b140eca9ba2743eac7594a54b4b8cd316bf7a28b154e2c9fa0fde43b90b7058

What do you think about putting in some of the additional checking and logging as used there to ensure we get a "good" value back from the function?

Comment (devkokov) [reply migration_steps/transform_casrec/transform/app/utilities/basic_data_table.py:24]: I think `get_source_table()` is used in just one place so I thought there's no point making `table_definition` an optional argument. If the key `source_table_name` doesn't exist in `table_definition` I think the code will bomb out with an error at that point, which is fine imo as all table definitions should have that key especially if the source table isn't defined anywhere else in the mapping.

**Review by townxelliot — APPROVED** [review]

**Pull request merged**